File: web/src/matrix.ts
// matrix.ts — pure-logic primitives shared by the v2 dashboard surfaces (matrix
// heatmap, coverage banner, table columns, drawer, worklist). No DOM, no React.
//
// The two-axis model (sub-project B): each account carries exposure_score (0–10,
// always) and impact_score (0–10 or null when Impact is Unknown — no BloodHound
// coverage). This module maps those scores to tiers, classifies coverage, counts
// accounts into the Exposure × Impact grid, and flags provisional (Unknown-impact)
// accounts. The design decisions referenced below are recorded in
// docs/superpowers/specs/2026-06-20-scoring-engine-v2-design.md:
//   D2 — impact_score null (or impact_known=false) ⇒ Impact Unknown;
//   D3 — coverage absent ⇒ "none";
//   D6 — tier cutoffs (≥8/≥6/≥4) mirror the Go engine exactly; Unknown is its own column.
import type { Account } from "./api"

export type Tier = "Critical" | "High" | "Medium" | "Low"
export const TIERS: Tier[] = ["Critical", "High", "Medium", "Low"]
export const IMPACT_UNKNOWN = "Unknown" as const

// axisTier mirrors B's per-axis cutoffs (>=8 Critical, >=6 High, >=4 Medium, else
// Low). This is documentation-only duplication of the 4 cutoffs (NOT the formula)
// from internal/risk/risk.go's tierOf; it is pinned by matrix.test.ts against the
// same boundary numbers as the Go golden tests so the two can't silently drift.
export function axisTier(v: number): Tier {
  if (v >= 8) return "Critical"
  if (v >= 6) return "High"
  if (v >= 4) return "Medium"
  return "Low"
}

// impactIsKnown: Impact is a usable number only when impact_known AND impact_score
// is non-null; both surfaces (provisional badge, matrix Unknown column) must use
// this so they can't drift.
export function impactIsKnown(a: Account): boolean {
  return a.impact_known === true && a.impact_score !== null
}

// isProvisional: true exactly when Impact is Unknown (level was derived from
// Exposure alone). The UI shows a "provisional" badge and never claims a number
// for Impact.
export function isProvisional(a: Account): boolean {
  return !impactIsKnown(a)
}

// coverageState: absent coverage (omitempty) means no enrichment record => "none".
export function coverageState(a: Account): "full" | "none" {
  return a.coverage === "full" ? "full" : "none"
}

export interface CoverageStats {
  enriched: number
  total: number
  partial: boolean
}
export function coverageStats(accts: Account[]): CoverageStats {
  const total = accts.length
  let enriched = 0
  for (const a of accts) if (coverageState(a) === "full") enriched++
  return { enriched, total, partial: total > 0 && enriched < total }
}

export type ImpactCol = Tier | typeof IMPACT_UNKNOWN
export const IMPACT_COLS: ImpactCol[] = [...TIERS, IMPACT_UNKNOWN]

export interface ExposureImpactMatrix {
  counts: Record<Tier, Record<ImpactCol, number>>
  total: number
  cell: (exp: Tier, imp: ImpactCol) => number
}

// exposureImpactMatrix: rows = Exposure tier, cols = Impact tier + an explicit
// Unknown column for impact_known=false accounts (which cannot be placed in an
// Impact tier). Tiers are derived from exposure_score / impact_score via axisTier.
export function exposureImpactMatrix(accts: Account[]): ExposureImpactMatrix {
  const counts = {} as Record<Tier, Record<ImpactCol, number>>
  for (const r of TIERS) {
    counts[r] = {} as Record<ImpactCol, number>
    for (const c of IMPACT_COLS) counts[r][c] = 0
  }
  let total = 0
  for (const a of accts) {
    const expT = axisTier(a.exposure_score)
    const impCol: ImpactCol =
      impactIsKnown(a) && a.impact_score !== null ? axisTier(a.impact_score) : IMPACT_UNKNOWN
    counts[expT][impCol]++
    total++
  }
  return { counts, total, cell: (exp, imp) => counts[exp][imp] }
}

// matrixMaxCount: the single largest cell count across the whole grid. Heatmap
// callers divide each cell by this to get a [0,1] intensity — kept here (tested)
// rather than computed in the component so the scale is pinned and never 0/NaN
// for an empty grid (returns 0, which callers guard before dividing).
export function matrixMaxCount(m: ExposureImpactMatrix): number {
  let max = 0
  for (const r of TIERS) for (const c of IMPACT_COLS) max = Math.max(max, m.counts[r][c])
  return max
}

File: internal/metrics/matrix.go
package metrics

import "github.com/watson0x90/PasswordAtTheDisco/internal/model"

// Tier is an axis tier (Exposure or Impact). ImpactUnknown is a column-only value
// for accounts with no BloodHound coverage (impact not a number).
type Tier string

const (
	TierCritical Tier = "Critical"
	TierHigh     Tier = "High"
	TierMedium   Tier = "Medium"
	TierLow      Tier = "Low"

	ImpactUnknown = "Unknown"
)

var tierRows = []Tier{TierCritical, TierHigh, TierMedium, TierLow}
var impactCols = []string{string(TierCritical), string(TierHigh), string(TierMedium), string(TierLow), ImpactUnknown}

// AxisTier mirrors internal/risk cutoffs and web/src/matrix.ts axisTier:
// >=8 Critical, >=6 High, >=4 Medium, else Low. Pinned by TestAxisTierBoundaries
// against the same boundary numbers as the Go engine and the TS mirror.
func AxisTier(v float64) Tier {
	switch {
	case v >= 8:
		return TierCritical
	case v >= 6:
		return TierHigh
	case v >= 4:
		return TierMedium
	default:
		return TierLow
	}
}

// levelMatrix is keyed [exposure tier][impact column] and is transcribed verbatim
// from web/src/matrix.ts LEVEL_MATRIX (itself derived from internal/risk levelMatrix).
// The Unknown column = the Exposure tier alone (provisional). Pinned by
// TestCellLevelMatchesMatrixTS.
var levelMatrix = map[Tier]map[string]Tier{
	TierCritical: {"Critical": TierCritical, "High": TierCritical, "Medium": TierHigh, "Low": TierMedium, ImpactUnknown: TierCritical},
	TierHigh:     {"Critical": TierCritical, "High": TierHigh, "Medium": TierHigh, "Low": TierMedium, ImpactUnknown: TierHigh},
	TierMedium:   {"Critical": TierCritical, "High": TierHigh, "Medium": TierMedium, "Low": TierLow, ImpactUnknown: TierMedium},
	TierLow:      {"Critical": TierHigh, "High": TierMedium, "Medium": TierMedium, "Low": TierLow, ImpactUnknown: TierLow},
}

// CellLevel returns the resulting risk Level for an (Exposure tier, Impact column).
func CellLevel(exp Tier, imp string) Tier { return levelMatrix[exp][imp] }

// Matrix is the Exposure×Impact distribution grid plus its largest cell (Max), used
// by the heatmap to normalize intensity.
type Matrix struct {
	Counts map[Tier]map[string]int `json:"counts"`
	Total  int                     `json:"total"`
	Max    int                     `json:"max"`
}

// impactColumn returns the Impact column for an account: a tier when Impact is a
// usable number (ImpactKnown AND ImpactScore non-nil), else ImpactUnknown. Mirrors
// web/src/matrix.ts impactIsKnown.
func impactColumn(a model.Account) string {
	if a.ImpactKnown && a.ImpactScore != nil {
		return string(AxisTier(*a.ImpactScore))
	}
	return ImpactUnknown
}

// BuildMatrix buckets accounts into the Exposure×Impact grid.
func BuildMatrix(accounts []model.Account) Matrix {
	m := Matrix{Counts: map[Tier]map[string]int{}, Total: 0}
	for _, r := range tierRows {
		m.Counts[r] = map[string]int{}
		for _, c := range impactCols {
			m.Counts[r][c] = 0
		}
	}
	for i := range accounts {
		a := accounts[i]
		m.Counts[AxisTier(a.ExposureScore)][impactColumn(a)]++
		m.Total++
	}
	for _, r := range tierRows {
		for _, c := range impactCols {
			if m.Counts[r][c] > m.Max {
				m.Max = m.Counts[r][c]
			}
		}
	}
	return m
}
